qiskit/transpiler/target.py

```diff
@@ -0,0 +1,652 @@
+                    qarg = (qarg,)
+                if inst in self._gate_map:
+                    if self.dt is not None:
+                        val.length = sched.duration * self.dt
+                    else:
+                        val.length = None
+                    if error_dict is not None:
+                        error_inst = error_dict.get(inst)
+                        if error_inst:
+                            error = error_inst.get(qarg)
+                            val.error = error
+                        else:
+                            val.error = None
+                    else:
+                        val.error = None
+                out_props[qarg] = val
+            if inst not in self._gate_map:
+                if inst_name_map is not None:
+                    self.add_instruction(inst_name_map[inst], out_props, name=inst)
+                else:
+                    raise ValueError(
+                        "An inst_name_map kwarg must be specified to add new "
+                        "instructions from an InstructionScheduleMap"
+                    )
+            else:
+                for qarg, prop in out_props.items():
+                    self.update_instruction_properties(inst, qarg, prop)
+
+    @property
+    def qargs(self):
+        """The set of qargs in the target."""
+        if None in self._qarg_gate_map:
+            return None
+        return set(self._qarg_gate_map)
+
+    def get_qargs(self, gate):
+        """Get the qargs for a given gate name
+
+        Args:
+           gate (str): The gate instance to get qargs for
+        Returns:
+            set: The set of qargs the gate instance applies to
+        """
+        if None in self._gate_map:
+            return None
+        return set(self._gate_map[gate])
+
+    def durations(self):
+        """Get an InstructionDurations object from the target
+
+        Returns:
+            InstructionDurations: The instruction duration represented in the
+                target
+        """
+        if self._instruction_durations is not None:
+            return self._instruction_durations
+        out_durations = []
+        for instruction, props_map in self._gate_map.items():
+            for qarg, properties in props_map.items():
+                if properties is not None and properties.length is not None:
+                    out_durations.append((instruction, list(qarg), properties.length, "s"))
+        self._instruction_durations = InstructionDurations(out_durations, dt=self.dt)
+        return self._instruction_durations
+
+    def timing_constraints(self):
+        """Get an :class:`~qiskit.transpiler.TimingConstraints` object from the target
+
+        Returns:
+            TimingConstraints: The timing constraints represented in the Target
+        """
+        return TimingConstraints(
+            self.granularity, self.min_length, self.pulse_alignment, self.aquire_alignment
+        )
+
+    def instruction_schedule_map(self):
```

Comment:
Currently the experiments depends on instmap so we cannot easily remove it for backward compatibility. Note that we should support V1 for sufficiently long deprecation period, so instmap is only portable container of instructions in the transition period. I've talked this with @mtreinish very carefully and we decided to keep the instmap and sync method (instmap -> instruction_properties) for now. Eventually they should be removed after deprecation period.

Reply:
We might drop `Target.instruction_durations` if we could add an alternative interface for retrieving the duration (in dt) of an Instruction (including PulseGate whose durations are currently stored in `InstructionScheduleMap`) to `Target`. It may be something like
```
backendv2.target["cx"][(0,1)].duration(unit="dt")
backendv2.target.duration("cx", (0,1), unit="dt")
```
(the former may require InstructionProperties class to be more than data class and the latter may require Target to have implementation very similar to `InstructionDurations`. I'm not sure those are good approaches.)
On the other hand, for `BackendV1` support, we need some functionality similar to `InstructionDuration` for now. So I think at least the deprecation of `InstructionDuration` should wait for the deprecation of BackendV1.